specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,438 @@
+* The Registration will take in a CLSID of the COM server as part of the Activator setup. The registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and abstract away the COM details from the developer.
+* The app will next query the ActivationEventArgs to check if the ActivationKind is of type Push
+* If the Activation is Push, the app will retrieve the IActivatedEventArgs and retrieve the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+* * It will subscribe to a Foreground In-memory Push event
+* * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler. 
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    std::wstring activatorGuid = L"{B8E58D9D-A229-44A8-B923-2823C3791BCE}";
+    InProcActivatorDetails details(activatorGuid);
+    PushNotificationChannelManager::RegisterPushNotificationActivator(details);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    auto args = AppLifecycle2::GetActivatedEventArgs();
+    auto kind = args.Kind2();
+
+    if (kind == ActivationKindExtension::Push)
+    {
+        auto activatedEventArgs = args.ActivatedArgs();
+        auto pushArgs = args.as<IPushNotificationActivatedEventArgs>();
+
+        // Call TakeDeferral to ensure that code runs in low power
+        pushArgs.GetDeferral();
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ActivationKindExtension::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushActivated([](const auto&, PushNotificationActivatedEventArgs args)
+        {
+                // Call TakeDeferral to ensure that code runs in low power
+                args.GetDeferral();
+
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+
+                // Call CompleteDeferral as good practise: Needed mainly for low power usage
+                args.CompleteDeferral();
+        });
+
+        auto remoteId = L"{5AB838FB-B64B-4FCD-B25F-B09781D56DD5}";
+        auto channelOperation = PushNotificationChannelManager::CreatePushNotificationChannelAsync(remoteId);
+
+        // Setup the inprogress event handler
+        channelOperation.Progress(
+            [](
+                IAsyncOperationWithProgress<PushNotificationChannelResult, PushNotificationChannelResult> const& /* sender */,
+                PushNotificationChannelResult const& args)
```

Comment:
For progress, Do we need to receive the PushNotificationChannelResult here? Can we just receive the Status from the implementation?

Reply:
The Status isn't sufficient if the developer wants to log the internal error code for debuggability. So returning a ChannelResult covers this scenario.